crypto/pem/pem_pkey.c

```diff
@@ -174,9 +178,13 @@ static EVP_PKEY *pem_read_bio_key_legacy(BIO *bp, EVP_PKEY **x,
             goto p8err;
         ret = ossl_d2i_PrivateKey_legacy(ameth->pkey_id, x, &p, len, libctx,
                                          propq);
-    } else if (selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) {
+    } else if (!(selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY)
+               && (selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY)) {
+        /* Trying legacy PUBKEY decoding only if we do not want private key. */
         ret = ossl_d2i_PUBKEY_legacy(x, &p, len);
-    } else if ((slen = ossl_pem_check_suffix(nm, "PARAMETERS")) > 0) {
+    } else if (!(selection & EVP_PKEY_KEYPAIR)
+               && (slen = ossl_pem_check_suffix(nm, "PARAMETERS")) > 0) {
```

Comment:
Should the "selection" logic here and above actually be written `(selection & WHATEVER) == 0` or `(selection & WHATEVER) != 0` as appropriate.

Reply:
Yeah, I'll change it.

Reply:
Fixed